### PR TITLE
Image pulling on the host

### PR DESCRIPTION
**Pull request by wedsonaf**
These are the changes in the kata runtime and agent that enable pulling images on the host and exposing layers as block devices to the VM, which are then mounted and overlayed to produce the final image. This layering allows sharing by containers with common layers, which is a desirable to reduce the memory footprint (and similar to what the overlay snapshotter does on the host). There is no dependency on image-rs.

The changes here are independent of which file system is used. Everything is driven by mount options, which can be used with any file system and snapshotter. (I believe this preferred over hard-coding file system names and functions in generic kata code.)

We have a reference snapshotter (tardev) that uses tarfs as the file system -- both are available here: https://github.com/kata-containers/tardev-snapshotter. We also have a fuse version of tarfs that we'll upload soon, as well as changes needed to enable dm-verity, the PRs for which will be uploaded once this is merged.

The kernel module is 664 LoC and has been tested in 5.15 and 6.1 kernels (the two latest LTSs). The fuse version is written in Rust and is currently 327 LoC.  The tardev snapshotter is written in Rust and totals 1095 LoC. It does not depend on virtiofs at all; it downloads layers, decompresses them, and appends an index to allow for efficient traversal by the fs impls. (In a subsequent PR, it will also append a dm-verity hash tree.)

This is the initial implementation of what I described in the Jan 26th meeting.

### CODE REVIEW

**Comment by wedsonaf** [comment]
Updated commit messages to mention https://github.com/confidential-containers/community/issues/66

**Comment by wedsonaf** [comment]
Removed usage of `strings.CutPrefix` as it's only available in go 1.20+.

**Comment by fitzthum** [comment]
@wedsonaf how does this relate to https://github.com/confidential-containers/documentation/issues/125?

**Comment by wedsonaf** [comment]
> @wedsonaf how does this relate to [confidential-containers/documentation#125](https://github.com/confidential-containers/documentation/issues/125)?

It's an implementation of the design documented there.

That design is the same as what I proposed in January, except that they use a modified version of erofs (to support 512-byte data alignment, which is what tar files offer) for the file system and Nydus for the snapshotter. (See [this slide](https://docs.google.com/presentation/d/1PU1XizarZt22eiVlQ_ug3bjtViST3qlya9OZQPWl8IM/edit#slide=id.p4) from Gerry's presentation after our first round of discussions.)

The Nydus/erofs combination demoed last week does not satisfy our requirements at MS:
1. It's not ready. Ours is ready with integrity support (with dm-verity root hashes enforced by policy). This is the first PR; dm-verity support is a smaller PR than this one (136 lines added to kata-agent).
2. Dependency on virtiofs: neither tdx- nor sev-snp-enabled guest kernels support virtiofs. So we can't really depend on it. Our implementation uses block devices, which works with tdx and sev-snp-enabled guests.
3. Performance: based on what I could see in the recording of last week's demo, images are shared via virtiofs with the guest, then a loop device (/dev/loop6 in the demo) is created, then erofs is mounted on top of that. It gets even worse if we have to use 9p instead of virtiofs. That's a lot of (arguably unnecessary) components in the data path; in our implementation in this PR, tarfs is mounted directly on the block device.
4. The modified erofs implementation is in kernel 6.4, which isn't even released yet ([currently in rc6](https://lore.kernel.org/lkml/CAHk-=wgmbjQu3bVSa5JJhEymeQhkr7N3cjg-CoNLVituUnfW8w@mail.gmail.com/)). The first LTS version is expected towards the end of the year, even then it can't be immediately used because CSPs and distros need to test first. Given that CSPs don't want to touch non-LTS kernels, this would be available for wide usage only next year. At MS we currently use 5.15. In any case, our tarfs module works on 5.15+ kernels (haven't really tested earlier kernels, but it doesn't depend on any fancy features, so I expect it to work) and is quite simple: we actually _want_ small and simple components in our TCB. For comparison, erofs is 9,040 LoC in 6.4 while tarfs is 664 LoC -- this is mostly due to features that we don't or can't use.
5. Dependency on Nydus: nydus is great but it's also big and complicated, it has 20,757 LoC in non-test files and is written in golang. Our tardev-snapshotter is written in Rust (which generally consumes less memory and is more efficient because it doesn't have a GC) and has 1,095 LoC -- it is much simpler. While it's not part of the confidential computing TCB, it is in the control plane and interacts with external untrusted components (e.g., container registries) so it's also preferable to have a reduced attack surface.

I'd also like to mention that we want containers that have common layers to be able to share them when running on the same pod. So we share individual layers as block devices. It's unclear to me what Nydus does: in the demo I didn't see individual layers. Anyway, sharing the layers separately also allows us to have combinations of encrypted/unencrypted layers -- this isn't possible if we squash all the layers.

Lastly, since our implementation doesn't depend on image-rs, it allows us to shave 8.6MB off our kata-agent binary (when pulling from the guest is not needed). The sizes of kata-agent are as follows for a release build: 31.9MB with image-rs, 23.3MB without image-rs. So we save on guest memory and reduce the amount of code in our TCB.

Having said all of the above, this PR does not preclude nydus/erofs, quite the contrary: it adds generic support for block-based layers to kata, which can be leveraged by other snapshotter/filesystems. It also doesn't touch any of the existing guest-pulling code -- it remains available for use if desired. This PR is also compatible with a proposal I posted on slack for supporting peer pods with a different snapshotter (that downloads layers to CSP disks) and a remote hypervisor that knows how to hot-add CSP disks to peer VMs.

This PR introduces 151 new lines of code, so it's quite simple and low risk.

It also allows us to have a working and simple image management implementation that works without a forked containerd in CC 0.7.0, and that can be upstreamed to main kata.

**Comment by wedsonaf** [comment]
Here is a demo of running a pod with tardev-snapshotter and tarfs:
[asciicast](https://asciinema.org/a/r8d9BEU9XnWRSgPZIUD7DVKpe)

**Comment by hsiangkao** [comment]
> 4\. For comparison, erofs is 9,040 LoC in 6.4 while tarfs is 664 LoC -- this is mostly due to features that we don't or can't use.

Personally I have no idea to reponse any of this (I don't think a read-only fs with more useful features are flaw, especially many features can be explicitly disabled at compile time or run time): guess what the current linux kernel LoC is? guess what the 2.6 kernel LoC is? Could you just duplicate all paths for your specific end-to-end use case?

The other points on my side are quite minor.

File: src/runtime/pkg/govmm/qemu/qmp.go
Comment (jepio) [inline src/runtime/pkg/govmm/qemu/qmp.go:825]: does qemu accept this? according to qemu protocol docs it really should be "host_device".

File: src/runtime/virtcontainers/kata_agent.go
Comment (jepio) [inline src/runtime/virtcontainers/kata_agent.go:1309]: why split out the return into two slices when both of them are appended to ctrStorages anyway?

File: src/runtime/virtcontainers/kata_agent.go
Comment (jepio) [inline src/runtime/virtcontainers/kata_agent.go:1314]: why are the args flipped compared to the usual convention - is the order somehow important?

**Comment by jepio** [comment]
This should target `main` instead of `CCv0`.

Do you have an example manifest that can be used to exercise the code paths in this PR?

**Comment by jepio** [comment]
As far as I can tell is not so much "image-pulling-on-host" as it is: "mount additional volumes from the host as block devices into the guest based on annotations".

So this is not coupled to any specific snapshotter and would be used in https://github.com/confidential-containers/documentation/issues/125 as well, because some of the modes rely on virtio-blk and not just virtiofs.

File: src/runtime/pkg/govmm/qemu/qmp.go
Comment (wedsonaf) [reply src/runtime/pkg/govmm/qemu/qmp.go:825]: Oh, this should be `host_device`. Thanks for catching this, will fix.

File: src/runtime/virtcontainers/kata_agent.go
Comment (wedsonaf) [reply src/runtime/virtcontainers/kata_agent.go:1309]: The layers must come before the rootfs because the rootfs is an overlay of the layers (otherwise mounting the rootfs fails). `handleBlkOCIMounts` returns the layers separately so that we can *prepend* them to `ctrStorages`, which makes them be mounted before the rootfs (which is the first entry in `ctrStorages`).

File: src/runtime/virtcontainers/kata_agent.go
Comment (wedsonaf) [reply src/runtime/virtcontainers/kata_agent.go:1314]: Because we want to *prepend* to `ctrStorages`. See my previous comment on why.

**Review by wedsonaf — COMMENTED** [review]
Thanks for the review @jepio!

File: src/runtime/virtcontainers/kata_agent.go
Comment (jepio) [reply src/runtime/virtcontainers/kata_agent.go:1309]: Very subtle - could you document this at line 1314

**Comment by wedsonaf** [comment]
> As far as I can tell this is not so much "image-pulling-on-host" as it is: "mount additional volumes from the host as block devices into the guest based on annotations".

Yes, "image pulling on the host" is the name of the PR only. The 3 commit messages just describe what the commits do.

> So this is not coupled to any specific snapshotter and would be used in [confidential-containers/documentation#125](https://github.com/confidential-containers/documentation/issues/125) as well, because some of the modes rely on virtio-blk and not just virtiofs.

That's right. I try to say this in one of my previous messages:  _this PR does not preclude nydus/erofs, quite the contrary: it adds generic support for block-based layers to kata, which can be leveraged by other snapshotter/filesystems._

File: src/runtime/virtcontainers/kata_agent.go
Comment (wedsonaf) [reply src/runtime/virtcontainers/kata_agent.go:1309]: Sure, will do.

**Comment by jepio** [comment]
> Regular manifests would work. To exercise this code path we need to configure containerd to use a snapshotter that returns mounts with these annotations, which is what tardev-snapshotter does.

From the code alone it's hard to figure out example annotations that need to be passed from the snapshotter. I'm not thinking about fully functional usage as container layers, but a manifest that results in host files showing up as block volumes in the guest to play around with.

File: src/runtime/virtcontainers/kata_agent.go
Comment (wedsonaf) [reply src/runtime/virtcontainers/kata_agent.go:1309]: Done in the latest version.

File: src/runtime/pkg/govmm/qemu/qmp.go
Comment (wedsonaf) [reply src/runtime/pkg/govmm/qemu/qmp.go]: Fixed in the latest version.

File: src/agent/src/mount.rs
Comment (c3d) [inline src/agent/src/mount.rs:325]: Maybe add a `.context("Creating work directory")` and similarly for next line, since there are multiple places that can fail in this function

**Comment by wedsonaf** [comment]
> > Regular manifests would work. To exercise this code path we need to configure containerd to use a snapshotter that returns mounts with these annotations, which is what tardev-snapshotter does.
> 
> From the code alone it's hard to figure out example annotations that need to be passed from the snapshotter. I'm not thinking about fully functional usage as container layers, but a manifest that results in host files showing up as block volumes in the guest to play around with.

I see. I've been using `crictl` to test this out. The [CRI api](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto#L212) doesn't support specifying mount options, so there is no obvious way to do this with `crictl` without a snapshotter.

I'm going to try out other tools to see if I can find one that allows me to specify mount options in a manifest or command line.

Meanwhile, you can find the annotations the snapshotter uses here: https://github.com/kata-containers/tardev-snapshotter/blob/main/snapshotter/src/snapshotter.rs#L121

File: src/agent/src/mount.rs
Comment (wedsonaf) [reply src/agent/src/mount.rs:325]: Good idea. Will do.

**Comment by wedsonaf** [comment]
@jepio -- I just added a 4-line commit to this PR that enables the following without a snapshotter:

```
$ sudo ./ctr run --rm -t --runtime io.containerd.kata.v2 --mount type=ext4,src=$PWD/disk.img,dst=/example,options=ro:io.katacontainers.fs-opt.block_device=file docker.io/library/busybox:latest example
/ # mount | grep example
/dev/vda on /example type ext4 (ro,relatime)
/ # ls /example
hello.txt   lost+found
/ #
```

Note that `$PWD/disk.img` is an ext4fs image I created previously. By adding the `io.katacontainers.fs-opt.block_device=file` option to the mount, the file is added as block device and is mounted into the pod with the right fs. Then it is bind-mounted into the container as you can see in the example above.

**Comment by jepio** [comment]
@wedsonaf the "Fixes:" check requires an issue in this repo, or an issue name that  starts with`github.com/kata-containers/...`. Can you create an issue and update the commit messages?

**Comment by ChengyuZhu6** [comment]
Hi, @wedsonaf . Maybe you could reserve some interfaces for image-rs  to support image-rs processing image blocks with encryption and integrity verification?

**Comment by ChengyuZhu6** [comment]
@wedsonaf This is the implementation of image block integrity check and mount in image-rs.  It can support all mount types, not just erofs tarfs.

**Comment by wedsonaf** [comment]
Rebased to `main` after speaking to @fidencio

**Comment by ChengyuZhu6** [comment]
hi @wedsonaf . It's the roadmap of image pulling with verity on the host. https://github.com/confidential-containers/confidential-containers/issues/142.
maybe we can discuss this ?

**Comment by fidencio** [comment]
I sincerely think this one is good to go as it's and it'll unblock us.

Very compotent people (@wedsonaf, @ChengyuZhu6, @jiangliu), way more compotent than I ever will be, are discussing how to build a solution atop of this, but I think this very first step can be unblocked.

There's an ongoing discussion on standardising how we pass the mount options from the runtime to the agent, and that discussion is happening as part of https://github.com/confidential-containers/confidential-containers/issues/142.

Regardless of my opinions on how that should be implemented, which has close to zero value considering that I'm not as familiar with this as the folks actively working on this, I that can be refactored in the coming weeks as part of the follow-up work.

I just cut Kata Containers 3.2.0-rc0, meaning that this work will **NOT** be part of an official release till 3.3.0, in a quarter from now, and we have that time in our favour to have that sorted out.

**Review by jiangliu — APPROVED** [review]

**Comment by fidencio** [comment]
> We do have different opinions about the message format

I'd like to have an explicit agreement from @wedsonaf, **before having this one merged**, that he agrees on continuing with the discussion and, **if needed**, adapting the code to the consensus reached in the coming discussions.

**Comment by ChengyuZhu6** [comment]
Thanks! @wedsonaf

**Comment by wedsonaf** [comment]
> > We do have different opinions about the message format
> 
> I'd like to have an explicit agreement from @wedsonaf, **before having this one merged**, that he agrees on continuing with the discussion and, **if needed**, adapting the code to the consensus reached in the coming discussions.

Yes, I'm happy to update this as we (the community) progress in our discussions.

**Comment by fidencio** [comment]
@wedsonaf, 

@wedsonaf, as our static checks are looking for an issue here, I've "cloned" the issue from the CoCo side to here .

With this in mind, please, change the reference in the commit messages from:
```
Fixes: confidential-containers/community#66
```

to:
```
Fixes: #7536
```

I'd have done it myself, but I have no authorisation to push to your branch. :-)

**Review by fidencio — APPROVED** [review]
lgtm, thanks @wedsonaf!

**Pull request merged**
